### PR TITLE
Add Support for Data Flow Task Launcher Sink

### LINKED ISSUE #21: Add Support for task-launcher-dataflow-sink
Currently this produces a [TaskLaunchRequest](https://github.com/spring-cloud/spring-cloud-task/blob/master/spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java) which is compatible with the platform specific (non SCDF) task-launcher sinks.  The dataflow task launcher requires the stream and launched task to be defined on a dataflow server and so the TLR is not appropriate if we are simply launching an existing task , e.g., TLR requires the uri to the task jar to be set, etc.  So we need to support both through properties configuration.

### LINKED ISSUE #22: Provide common command line args to the launch request.
Depends on #21,  The source can provide a core set of command line arguments related to the sftp context, server, filename, etc., along with a way to provide additional arguments.

### PR DESCRIPTION
**Pull request by dturanski**
Fixes #21 and Fixes #22

* Change `sftp.task-launcher-output` from boolean to enum (NONE, STANDALONE, DATAFLOW). `STANDALONE` behaves as before with `sftp.task-launcher-output=true`. `DATAFLOW` outputs a Task Launch Request compatible with the `dataflow-task-launcher-sink`. 

NOTE this is a breaking change, so more to follow:

* The last commit changes the `sftp.batch` property prefix to `sftp.task` and removes `job` from property names. This is appropriate since the `TaskLaunchRequest` is provided by Spring-Cloud-Task, which is independent of, but supports Spring Batch. Since this PR adds the `application-name` property, `task.application-name` is more appropriate than `batch.application-name` and the same is true for all properties pertaining to the TLR. 

* Adds README content pertaining to Task Launch Requests.   

* Fixes the `@ConditionalOnProperty` intention in `SftpSourceTaskLauncherConfiguration`. The annotation had no effect since the method was only `@Transformer` and not `@Bean`.  Having this work correctly is required(convenient at least) for providing multiple `sftp.task-launcher-output` transformations.

### CODE REVIEW

File: spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/tasklauncher/SftpSourceTaskLauncherConfiguration.java [hunk 1]
```diff
@@ -62,36 +68,62 @@
...
+			Assert.hasText(sftpSourceTaskProperties.getApplicationName(),
+				"'applicationName' is required for DataFlow Task Launcher.");
+		}
+	}
+
+	@Bean
+	@ConditionalOnProperty(name = "sftp.task-launcher-output", havingValue = "STANDALONE")
+	@IdempotentReceiver("idempotentReceiverInterceptor")
+	@ServiceActivator(inputChannel = "sftpFileTaskLaunchChannel", outputChannel = Source.OUTPUT)
+	public MessageProcessor<Message> sftpFileTaskLauncherTransformer() {
```
Comment (chrisjs) [inline spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/tasklauncher/SftpSourceTaskLauncherConfiguration.java]: should this be called `standaloneTaskLaunchRequestTransformer` similar to `dataflowTaskLaunchRequestTransformer`?

File: spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/tasklauncher/SftpSourceTaskLauncherConfiguration.java [hunk 2]
```diff
@@ -62,36 +68,62 @@
...
+		};
 	}
 
+
 	private Map<String, String> getEnvironmentProperties() {
 		Map<String, String> environmentProperties = new HashMap<>();
-		environmentProperties.put(DATASOURCE_URL_PROPERTY_KEY, sftpSourceBatchProperties.getDataSourceUrl());
-		environmentProperties.put(DATASOURCE_USERNAME_PROPERTY_KEY, sftpSourceBatchProperties.getDataSourceUserName());
+		environmentProperties.put(DATASOURCE_URL_PROPERTY_KEY, sftpSourceTaskProperties.getDataSourceUrl());
+		environmentProperties.put(DATASOURCE_USERNAME_PROPERTY_KEY, sftpSourceTaskProperties.getDataSourceUserName());
```
Comment (chrisjs) [inline spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/tasklauncher/SftpSourceTaskLauncherConfiguration.java:122]: not part of this PR/issue but since its minor and we are in here, looks like we overlooked adding the datasource password ("sa" on localserver has no password..)

**Review by chrisjs — COMMENTED** [review]
LGTM with a couple minor requests

**Comment by chrisjs** [comment]
two things i can across while testing so far:

* when doing an app info the default type for `sftp.task-launcher-output` shows: `<none>` -- does that represent "no default value defined" or is it the rendering of `NONE` as shown in the description: `Default is `NONE` 

* when trying to define a stream with the following definition, the task launcher fails to launch the task:

```
app register --name sftp --type source --uri file:////Users/cschaefer/work/sftp/apps/sftp-source-kafka/target/sftp-source-kafka-2.0.3.BUILD-SNAPSHOT.jar
app register --name log --type sink --uri maven://org.springframework.cloud.stream.app:log-sink-kafka:2.0.1.RELEASE
app register --name task-launcher-local --type sink --uri maven://org.springframework.cloud.stream.app:task-launcher-local-sink-kafka:2.0.1.RELEASE

stream create --name inboundSftp --definition "sftp --username=user --password=pass --host=127.0.0.1 --port=6666 --allow-unknown-keys=true --task-launcher-output=DATAFLOW --remote-dir=/tmp/blah --resource-uri=file:////Users/cschaefer/work/dataflow/spring-cloud-dataflow-samples/batch/file-ingest/target/ingest-1.0.0.jar --data-source-url=jdbc:h2:tcp://localhost:19092/mem:dataflow --data-source-user-name=sa --local-file-path-parameter-value=/tmp/ --application-name=sftpTest | task-launcher-local" --deploy
```

with the following exception snippet:

```
2018-08-20 16:09:42.881  INFO 98648 --- [container-0-C-1] o.s.c.task.launcher.TaskLauncherSink     : Launching Task for the following uri null
2018-08-20 16:09:42.887 ERROR 98648 --- [container-0-C-1] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessageHandlingException: nested exception is org.springframework.cloud.deployer.resource.support.ResourceNotResolvedException, failedMessage=GenericMessage [payload=byte[115], headers={sequenceNumber=1, file_remoteDirectory=/tmp/blah/, sequenceSize=1, deliveryAttempt=3, kafka_timestampType=CREATE_TIME, kafka_receivedMessageKey=null, kafka_receivedTopic=inboundSftp.sftp, kafka_offset=0, scst_nativeHeadersPresent=true, kafka_consumer=org.apache.kafka.clients.consumer.KafkaConsumer@128f1765, correlationId=b1b1a896-951f-e71f-2c76-1191ce89e31f, kafka_receivedPartitionId=0, contentType=application/json, kafka_receivedTimestamp=1534795779713}]
```

seems the `resource-uri` isn't getting passed through (or my stream def is just wrong)?

**Comment by dturanski** [comment]
Yeah, app-info always does that for enum AFAICT.

**Comment by dturanski** [comment]
`DATAFLOW` will only work with task-launcher-dataflow (which is also in PR).  You need `STANDALONE` for the existing ones.  That's correct, uri is ignored for `DATAFLOW` as well as any datasource properties. Basically a different payload type.

**Comment by chrisjs** [comment]
ah right, thanks. worked for me, merged in d498a0
